inline-c-cpp/src/Language/C/Inline/Cpp/Exception.hs

```diff
@@ -0,0 +1,239 @@
+-- | A module that contains exception-safe equivalents of @inline-c@ QuasiQuoters.
+
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Language.C.Inline.Cpp.Exception
+  ( CppException(..)
+  , CppExceptionPtr
+  , toSomeException
+  , throwBlock
+  , tryBlock
+  , catchBlock
+  ) where
+
+import           Control.Exception.Safe
+import qualified Data.ByteString.Unsafe as BS (unsafePackMallocCString)
+import           Data.ByteString (ByteString)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Data.Text.Encoding.Error as T
+import qualified Language.C.Inline as C
+import qualified Language.C.Inline.Internal as C
+import qualified Language.C.Inline.Cpp as Cpp
+import           Language.C.Inline.Cpp (AbstractCppExceptionPtr)
+import           Language.Haskell.TH
+import           Language.Haskell.TH.Quote
+import           Foreign
+import           Foreign.C
+import           System.IO.Unsafe(unsafePerformIO)
+
+C.context Cpp.cppCtx
+C.include "HaskellException.hxx"
+
+-- | An exception thrown in C++ code.
+data CppException
+  = CppStdException CppExceptionPtr ByteString (Maybe ByteString)
+  | CppHaskellException SomeException
+  | CppNonStdException CppExceptionPtr (Maybe ByteString)
+
+instance Show CppException where
+  showsPrec p (CppStdException _ msg typ) = showParen (p >= 11) (showString "CppStdException e " . showsPrec 11 msg . showsPrec 11 typ)
```

Comment:
@roberth Was it intended that there's this ` e ` in the message?

Reply:
Ah, now I get it. The `e` is suppoed to represent the `_` (first field of constructor `CppStdException`) which is not `show`n because it's a pointer.